Add GPU implementation of SparseSliceGrad

benbarsdell opened the pull request:
cc @nluehr

benbarsdell committed:
Enable GPU tests in SparseSliceOpTest

- Also fixes formatting of test data.

benbarsdell committed:
Add GPU implementation of SparseSliceGrad

reedwm commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -359,13 +358,12 @@ __device__ OutType upper_bound(const T* first, OutType count, T val) {
-  const T* orig = first;
-  const T* it = nullptr;
+template <typename T, typename OutType = int32, typename Iterator = const T*>
+__device__ OutType lower_bound(Iterator first, OutType count, T val) {
```
Add a comment, since it's a bit unclear what this does. E.g., "Returns the first element in [first, first+count) that is not less than "val". Assumes [first, first+count) is sorted."

Also similarly for upper_bound.

benbarsdell replied:
Done.

reedwm commented on tensorflow/python/kernel_tests/sparse_ops/sparse_slice_op_test.py:
```diff
@@ -284,39 +284,45 @@ def testGradients(self):
+                                          0, 23, 0,
+                                          0, 0, 33, 0])
+    # pyformat: enable
+
```
I think it would be worthwhile to test a case where the input is non-empty, but the output is empty. E.g., using `_SparseTensor_4x6` and `start_size = [2, 1], [2, 1]`.

benbarsdell replied:
Done.

reedwm commented on tensorflow/core/kernels/sparse_slice_grad_op_gpu.cu.cc:
```diff
@@ -0,0 +1,167 @@
+                                      const T* backprop_val_grad, T* val_grad) {
+  for (int64 input_nz : GpuGridRangeX<int64>(input_nnz)) {
+    // Search for the input index in the output indices.
+    // Note: It would be faster to directly test if the input index is within
+    // the slice volume (and also to use flattened indexes), but unfortunately
+    // we don't have the dense shapes so we can't do that.
```
But even if you had the dense shapes, you'd still need to call lower_bound to do the binary search, because you need to find the index `output_nz` so you can copy the correct value from `backprop_val_grad` to `val_grad[input_nz]`

benbarsdell replied:
That's correct; the comment is just pointing out that the test for being inside the slice volume could be done without the need to look up `output_indices[output_nz]`, and that `lower_bound()` would only need to be called inside the true branch.
I've added the word "first" to make this clearer.

benbarsdell committed:
Minor improvements for SparseSliceGrad GPU implem

- Tweaks comment wording.
- Passes nullptr for unused parameter.
- Adds comments describing lower/upper_bound.
- Adds test for empty output from non-empty input.

benbarsdell commented on tensorflow/core/kernels/sparse_slice_grad_op_gpu.cu.cc:
```diff
@@ -0,0 +1,167 @@
+    MultiIndexSearchIterator input_indices_iter(
+        IndexIterator(int64_t(0)),
+        MultiIndexSearchFunctor(rank, input_indices_mat.data(),
+                                input_start_flat.data()));
```
Good catch, done.
